designs/package_during_deploy.md

```diff
@@ -0,0 +1,175 @@
+Package during `sam deploy`
+====================================
+
+
+What is the problem?
+--------------------
+
+Today while using `sam deploy` the specified `--template-file` is expected to have packaged artifacts references in the given template file.
+
+This is driven by the following workflow.
+
+`sam build` -> `sam package --s3-bucket ... --s3-prefix ... --output-template-file packaged.yaml`.
+
+This workflow builds the requisite code and packages those built artifacts into an s3 bucket, optionally under a given s3 prefix.
+
+If a developer can optionally cut through this process without requiring an explicit `package` command, but rather have `sam deploy` package to a given s3 bucket, it cuts the number of steps before needing to deploy and test in the cloud.
+
+This also reduces friction in the `author` and `test` loop.
+
+Ideal end state.
+
+`sam build` -> `sam deploy ..... --s3-bucket ..`
+
+
+What will be changed?
+---------------------
+
+Addition of extra parameters that are currently supported by `sam package` over to `sam deploy` with all of them being optional.
+
+Additional parameters that need to be on `sam deploy` that are not on `sam package`.
+
+* `--metadata`
+* `--use-json`
+
+Parameters that dont need to be added.
+
+* `--output-template-file`
+ * 	An explicit `output-template-file` is created on the fly during packaging in the deploy phase.
+
+If the expectation is to package and deploy in one command, One can now do.
+
+`sam deploy --stack-name sam-package-on-deploy --capabilities CAPABILITY_IAM --s3-bucket sam-package-bucket`
+
+There is no explicit need to pass in a `--template-file` parameter, if one is not passed in it to defaults to trying to find the template.yaml that was generated by `sam build` which is located at `.aws-sam/build/template.yaml`
```

Comment:
If `sam build` was never run what happens? Do we pick up the default template from the cwd or fail? All of our other commands pick up the default but could see a case to fail and leave the door open for a deploy that builds, packages, and deploys. Would be worth describing what will happen here.

Reply:
yeah, we pick up the default template. Can describe that case.